test/functional/treesitter/parser_spec.lua

```diff
@@ -102,6 +102,20 @@ void ui_refresh(void)
   }
 }]]
 
+  it('get_parser() changes the language of the parser', function()
+    local res = exec_lua([[
+      parser = vim.treesitter.get_parser(0, "c")
+      return parser:lang()
+    ]])
+    eq("c", res)
+
+    local ok, err = unpack(exec_lua([[
+       return {pcall(vim.treesitter.get_parser, 0, "cpp")}
+    ]]))
+    assert(not ok)
+    assert(string.find(err, "no parser for 'cpp' language", 1, true))
+  end)
+
```

Comment:
I don't think this tests actually tests your changes, why don't you change the filetype of the buffer in here?

Reply:
the test fails without this change, get_parser reads from the file type if the lang isn't given https://github.com/neovim/neovim/blob/3521bf7672d2a33e60e2656043374a1584a0285b/runtime/lua/vim/treesitter.lua#L85-L87, the bug here is just when requesting a different language for the same buffer. I can change it if needed, but looks like it's the same.

Reply:
Then I'd like to see the error message changed, at least to include the buffer number.

Reply:
I see what you are trying to do now. The test will need to be changed to actually check what we want.
The issue I see now is that when one guy requests a parser of language A, then another guy requests a parser for language B, then the A parser will be lost.

Reply:
> The issue I see now is that when one guy requests a parser of language A, then another guy requests a parser for language B, then the A parser will be lost.

but then is it okay to have different parsers for the same buffer? Should we cache per-lang and per-buffer instead of just per-buffer?

Reply:
It is okay, but my problem is that we explicitly call destroy on the previously cached parser, which IMO is a big no no.